source/tutorials/batch-inference/csv-batch-inference.rst

```diff
@@ -0,0 +1,160 @@
+
+Edit the generated ``valohai.yaml`` so that it looks like this:
+
+.. code-block:: yaml
+
+    ---
+
+    - step:
+        name: Batch Inference
+        image: tensorflow/tensorflow:2.5.0
+        command:
+        - pip install pandas valohai-utils
+        - python batch_inference.py
+        inputs:
+        - name: model
+          default: s3://valohai-public-files/tutorials/batch-inference/csv-batch-inference/model.zip
+        - name: images
+          default: s3://valohai-public-files/tutorials/batch-inference/csv-batch-inference/data.csv
+
+What we are doing here is defining a single step for our machine learning pipeline, which is the Batch Inference step. We will run on top of the official ``tensorflow/tensorflow:2.5.0`` Docker image, first install the ``valohai-utils`` Python library and then run our batch inference code.
+
+Let's test that everything is set up correctly by running on Valohai:
+
+.. code-block:: bash
+
+    vh exec run --adhoc "Batch Inference"
+
+If everything went as planned, we should see our Valohai execution end after finding out that ``batch_inference.py`` is missing:
+
+.. image:: batch-inference-tutorial-1.png
+   :alt: Error, but success!
+
+Unpacking the Model
+--------------------
+
+Today we are unpacking the model ourselves. Let's get started by creating and opening up ``batch_inference.py`` in your favorite editor!
+
+Add these imports to the beginning of the file:
+
+.. code-block:: python
+
+    import json
+    from zipfile import ZipFile
+
+    import pandas as pd
+    import tensorflow as tf
+    import valohai as vh
+
+For unpacking the model, we will only need `zipfile` and `valohai`, but we will use the rest of the imports soon enough.
+
+Next, unpack the model to a folder called model in the current working directory:
+
+.. code-block:: python
+
+    with ZipFile(vh.inputs('model').path(process_archives=False), 'r') as f:
+        f.extractall('model')
+
+Done!
+
+Loading and Using Our Model
+---------------------------
+
+Begin by loading our model:
+
+.. code-block:: python
+
+    model = tf.keras.models.load_model('model')
+
+Easy, huh? Let's load up the data:
+
+.. code-block:: python
+
+    csv = pd.read_csv('data.csv')
+    labels = csv.pop('target')
+    data = tf.data.Dataset.from_tensor_slices((dict(csv), labels))
+    batch_data = data.batch(batch_size=32)
+
+Aaand we are almost done. Run the model with the loaded up data. While we're at it, let's log and save the results as a JSON file:
+
+.. code-block:: python
+
+    results = model.predict(batch_data)
+
+    values = [str(nums[0]) for nums in results]
+    metadata = {k:v for (k, v) in zip(range(1, len(values) + 1), values)}
+
+    for _, value in metadata.items():
+        with vh.logger() as logger:
+            logger.log("result", value)
+
+    with open(vh.outputs().path('results.json'), 'w') as f:
+        json.dump(metadata, f)
```

Comment:
Perhaps we should change `Output.path()` to return a real `pathlib.Path`, so you could use [`Path.write_text()`](https://docs.python.org/3.9/library/pathlib.html#pathlib.Path.write_text) to make this

```python
    vh.outputs().path('results.json').write_text(json.dumps(metadata))
```